scrapy_poet/commands.py

```diff
@@ -0,0 +1,115 @@
+import datetime
+from pathlib import Path
+from typing import Type
+
+import andi
+import scrapy
+from freezegun import freeze_time
+from scrapy import Request
+from scrapy.commands import ScrapyCommand
+from scrapy.crawler import Crawler
+from scrapy.exceptions import UsageError
+from scrapy.http import Response
+from scrapy.utils.misc import load_object
+from twisted.internet.defer import inlineCallbacks
+from web_poet import ItemPage
+from web_poet.testing import save_fixture
+
+from scrapy_poet import callback_for
+from scrapy_poet.downloadermiddlewares import DEFAULT_PROVIDERS, InjectionMiddleware
+from scrapy_poet.injection import Injector
+
+saved_dependencies = []
+saved_items = []
+
+
+class SavingInjector(Injector):
+    @inlineCallbacks
+    def build_instances_from_providers(
+        self, request: Request, response: Response, plan: andi.Plan
+    ):
+        instances = yield super().build_instances_from_providers(
+            request, response, plan
+        )
+        saved_dependencies.extend(instances.values())
+        return instances
+
+
+class SavingPipeline:
+    def process_item(self, item, spider):
+        saved_items.append(item)
+        return item
+
+
+class SavingInjectionMiddleware(InjectionMiddleware):
+    def __init__(self, crawler: Crawler) -> None:
+        super().__init__(crawler)
+        self.injector = SavingInjector(
+            crawler,
+            default_providers=DEFAULT_PROVIDERS,
+            overrides_registry=self.overrides_registry,
+        )
+
+
+def spider_for(injectable: Type[ItemPage]) -> Type[scrapy.Spider]:
+    class InjectableSpider(scrapy.Spider):
+        name = "injectable"
+
+        url = None
+        custom_settings = {
+            "SCRAPY_POET_PROVIDERS": DEFAULT_PROVIDERS,
```

Comment:
Are we preventing custom providers on purpose?

Reply:
No, this is one of the parts where I'm not sure I'm supporting various customizations well.

Reply:
(this code was combined from existing tests code and some new tests code from the unmerged Kevin's PR and then heavily trimmed down so something was lost or kept where it shouldn't be)

Reply:
We'd need to support both `DEFAULT_PROVIDERS` and `SCRAPY_POET_PROVIDERS`, similar to https://github.com/scrapinghub/scrapy-poet/blob/master/scrapy_poet/injection.py#L53-L57

Reply:
I think we can just remove this line so that the project settings are used?

Reply:
Yes, that works.